Move EDS Clusters initialization after ADS

ramaraochavali opened the pull request:
Fixes https://github.com/envoyproxy/envoy/issues/2759

htuch commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -238,38 +263,13 @@ ClusterManagerImpl::ClusterManagerImpl(const envoy::config::bootstrap::v2::Boots
     }
   }
 
-  Optional<std::string> local_cluster_name;
```
Why did this need to move?

htuch commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -238,38 +263,13 @@ ClusterManagerImpl::ClusterManagerImpl(const envoy::config::bootstrap::v2::Boots
   });
 
-  // Now setup ADS if needed, this might rely on a primary cluster and the
-  // thread local cluster manager.
```
This comment was there to point out that ADS init must happen after the above `tls_->set`. We need TLS to be able to initialize gRPC async client manager bits. I'm surprised the tests pass without this (specifically `ads_integration_test` with Google gRPC).

htuch replied:
Actually, I think the need for this went away when I fixed the stats scope to be created on the main thread. Feel free to ignore this comment, if an issue arises I can go and fix it, since this is `Grpc::AsyncClient` implementation stuff.

ramaraochavali commented:
@htuch addressed the feedback and added a `ConfigIntegrationTest` that validates this case. I verified that with old code that test fails and new code it passes. PTAL.

htuch commented on test/integration/ads_integration_test.cc:
```diff
@@ -434,5 +435,58 @@ TEST_P(AdsFailIntegrationTest, ConnectDisconnect) {
+      ads_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+      ads_cluster->set_name("ads_cluster");
+
+      // Add EDS Cluster that uses ADS as config Source.
```
Nit: s/Add EDS/Add EDS static/

htuch commented on test/integration/ads_integration_test.cc:
```diff
@@ -434,5 +435,58 @@ TEST_P(AdsFailIntegrationTest, ConnectDisconnect) {
+      ads_eds_cluster->set_type(envoy::api::v2::Cluster::EDS);
+      auto* eds_cluster_config = ads_eds_cluster->mutable_eds_cluster_config();
+      auto* eds_config = eds_cluster_config->mutable_eds_config();
+      eds_config->mutable_ads()->InitAsDefaultInstance();
```
Nit: you can just write `eds_config->mutable_ads();`

htuch commented on test/integration/ads_integration_test.cc:
```diff
@@ -231,6 +231,7 @@ class AdsIntegrationTest : public HttpIntegrationTest, public Grpc::GrpcClientIn
             TestEnvironment::runfilesPath("test/config/integration/certs/upstreamcacert.pem"));
       }
     });
+
```
Nit: superfluous whitespace change.

htuch commented on test/integration/ads_integration_test.cc:
```diff
@@ -434,5 +435,58 @@ TEST_P(AdsFailIntegrationTest, ConnectDisconnect) {
+INSTANTIATE_TEST_CASE_P(IpVersionsClientType, AdsConfigIntegrationTest,
+                        GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// Validate that we don't crash on EDS Cluster that uses ADS as ConfigSource.
```
More nits: don't crash on EDS static cluster...

This is a regression test for <GitHub issue link>.

htuch commented on test/integration/ads_integration_test.cc:
```diff
@@ -434,5 +435,58 @@ TEST_P(AdsFailIntegrationTest, ConnectDisconnect) {
+  ads_connection_ = fake_upstreams_[1]->waitForHttpConnection(*dispatcher_);
+  ads_stream_ = ads_connection_->waitForNewStream(*dispatcher_);
+  ads_stream_->startGrpcStream();
+  ads_stream_->finishGrpcStream(Grpc::Status::Ok);
```
Do you need all of this just to verify initialization?

ramaraochavali replied:
with out this, some ipv6 tests are failing.

htuch replied:
This seems surprising. If you reduce it to `initialize()` only, it fails on IPv6 abut not IPv4?

ramaraochavali replied:
Yes.

ramaraochavali replied:
I tried reducing it to waitForHttpConnection then ipv6 test failed. If I reduce to initialise the test is passes with old code also. That’s why added these additional steps

htuch commented on test/integration/ads_integration_test.cc:
```diff
@@ -434,5 +435,58 @@ TEST_P(AdsFailIntegrationTest, ConnectDisconnect) {
   ads_stream_->finishGrpcStream(Grpc::Status::Internal);
 }
 
+class AdsConfigIntegrationTest : public HttpIntegrationTest,
```
I'm wondering if this could have been also done as a server unit test, i.e. just verifying the config loads. I think it's fine to do this here, but I'd also try and minify the amount of stuff we have; all we want to do is verify initialization with an ADS configuration. We don't care about setting up an ADS server or actually serving up ADS.

ramaraochavali replied:
i agree. I thought about it being a unit test. I added it here so that we could add ADS related config tests that ensure ADS connection works with config changes in future if needed.
